consul/base_client.go

```diff
@@ -0,0 +1,190 @@
+package consul
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/solo-io/gloo-api/pkg/api/types/v1"
+	"github.com/solo-io/gloo-storage"
+	"github.com/solo-io/gloo/pkg/log"
+)
+
+// TODO: evaluate efficiency of LSing a whole dir on every op
+// so far this is preferable to caring what files are named
+type baseClient struct {
+	rootPath      string
+	consul        *api.Client
+	syncFrequency time.Duration
+}
+
+type ConfigObjectType string
+
+const (
+	configObjectTypeUpstream    = "Upstream"
+	configObjectTypeVirtualHost = "VirtualHost"
+)
+
+func (c *baseClient) Create(item v1.ConfigObject, t ConfigObjectType) (v1.ConfigObject, error) {
+	p, err := toKVPair(c.rootPath, item)
+	if err != nil {
+		return nil, errors.Wrapf(err, "converting %s to kv pair", item.GetName())
+	}
+
+	// error if the key already exists
+	existingP, _, err := c.consul.KV().Get(p.Key, &api.QueryOptions{RequireConsistent: true})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to query consul")
+	}
+	if existingP != nil {
+		return nil, storage.NewAlreadyExistsErr(
+			errors.Errorf("key found for configObject %s: %s", item.GetName(), p.Key))
+	}
+	if _, err := c.consul.KV().Put(p, nil); err != nil {
+		return nil, errors.Wrapf(err, "writing kv pair %s", p.Key)
+	}
+	// set the resource version from the CreateIndex of the created kv pair
+	p, _, err = c.consul.KV().Get(p.Key, &api.QueryOptions{RequireConsistent: true})
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting newly created kv pair %s", p.Key)
+	}
+	// set resourceversion on clone
+	configObjectClone := proto.Clone(item).(v1.ConfigObject)
```

Comment:
there might be a race between the put() and ge() above that might result in get returning a more updated version of the object; such that just cloning item and updating its version may not be enough;

consider using the returned object from get and not use item